### PR TITLE
fix(medusa): Rounding issues on line item adjustments

### PR DESCRIPTION
**Pull request by adrien2p**
**What**
Depending on the size of the cart and the amount of items and their price amount the total. The discount displayed can end up having rounding issues.

The rounding issue is a know subject world wide for which there is no concrete way to get around it. It is widely accepted, but even though, the purpose of that pr is to reduce the probability that a rounding issue occurs during the discount calculation in the end result.

FIXES CORE-1281

### CODE REVIEW

File: packages/medusa/src/services/discount.ts [hunk 1]
```diff
@@ -673,14 +675,38 @@ class DiscountService extends TransactionBaseService {
...
+        adjustment = Number(
+          Math.trunc(adjustment) + stringifiedAdjustment.slice(pointIndex + 1)
+        )
+
+        const zeroCount = stringifiedAdjustment.slice(pointIndex + 1).length
+        multiplierFactor = Number(
+          new Array(zeroCount).fill("0").reduce(function (acc, v) {
+            return acc + v
+          }, "1")
+        )
```
Comment (carlos-r-l-rodrigues) [inline packages/medusa/src/services/discount.ts:695]: Probably it is a good idea to create a helper function for it.

`pointIndex` can be -1 if there is no "." in it.

Comment (adrien2p) [reply packages/medusa/src/services/discount.ts:695]: we are doing it only if !== -1. What do you mean?

Comment (adrien2p) [reply packages/medusa/src/services/discount.ts]: realised that I did not pushed, but we will close the pr in order to think it through. Simple changes does not work as we expected it to arrived and will have other impact that we don't want to rush into too quickly

**Review by carlos-r-l-rodrigues — APPROVED** [review]
LGTM 🚀

File: integration-tests/api/__tests__/admin/draft-order/draft-order.js [hunk 2]
```diff
@@ -585,7 +582,7 @@ describe("/admin/draft-orders", () => {
           adjustments: expect.arrayContaining([
             expect.objectContaining({
               item_id: lineItem1.id,
-              amount: discountAmount * lineItem1WeightInTotal,
+              amount: 444.4444444444444, // discountAmount * (line item amount / amount),
```
Comment (carlos-r-l-rodrigues) [inline integration-tests/api/__tests__/admin/draft-order/draft-order.js:585]: non-blocking: maybe we could use [expect.closeTo](https://jestjs.io/docs/expect#expectclosetonumber-numdigits) if our jest version supports it

Comment (adrien2p) [reply integration-tests/api/__tests__/admin/draft-order/draft-order.js:585]: that is nice,  did not know about this one 😂

Comment (adrien2p) [reply integration-tests/api/__tests__/admin/draft-order/draft-order.js:585]: There is no close to when expecting here. Only at the higher level `expect(x).toBeCloseTo`

Comment (carlos-r-l-rodrigues) [reply integration-tests/api/__tests__/admin/draft-order/draft-order.js:585]: probably the version we're using doesn't support it yet

Comment (adrien2p) [reply integration-tests/api/__tests__/admin/draft-order/draft-order.js]: I exctrated it at the top level when needed 👍

File: packages/medusa/src/api/middlewares/transform-query.ts [hunk 3]
```diff
@@ -12,12 +12,21 @@ import { FindConfig, QueryConfig, RequestQueryFields } from "../../types/common"
 import { omit } from "lodash"
```
Comment (olivermrbl) [inline packages/medusa/src/api/middlewares/transform-query.ts:12]: question: What happened in this file? 😅

File: packages/medusa/src/services/discount.ts [hunk 4]
```diff
@@ -620,6 +620,7 @@ class DiscountService extends TransactionBaseService {
       let adjustment = 0
 
       if (!lineItem.allow_discounts) {
+        // return { adjustment, multiplierFactor: 0 }
```
Comment (olivermrbl) [inline packages/medusa/src/services/discount.ts:623]: todo: rm

File: packages/medusa/src/services/discount.ts [hunk 5]
```diff
@@ -671,14 +675,13 @@ class DiscountService extends TransactionBaseService {
...
+
+        adjustment = nominator * totalItemPercentage
       } else {
         adjustment = value * lineItem.quantity
       }
 
-      // if the amount of the discount exceeds the total price of the item,
-      // we return the total item price, else the fixed amount
-      return adjustment >= fullItemPrice ? fullItemPrice : adjustment
+      return Math.min(adjustment, fullItemPrice)
```
Comment (olivermrbl) [inline packages/medusa/src/services/discount.ts:684]: praise: Haha, nice one!

**Review by olivermrbl — APPROVED** [review]
LGTM - great work @adrien2p! 💪 

As next steps, I suggest we ask them to test a snapshot version of this one. I'll publish one and throw it their way. Cool with you? 💯

**Comment by adrien2p** [comment]
> LGTM - great work @adrien2p! 💪 
> 
> 
> 
> As next steps, I suggest we ask them to test a snapshot version of this one. I'll publish one and throw it their way. Cool with you? 💯 

Totally fine for me, i prefer doing it that way for such fixes 👌

File: packages/medusa/src/api/middlewares/transform-query.ts (hunk 3 above)
Comment (adrien2p) [reply packages/medusa/src/api/middlewares/transform-query.ts:12]: I have ported the update from develop to apply it in the handler 😅

**Review by fPolic — APPROVED** [review]
LGTM 💪

**Comment by srindom** [comment]
Think this is safe to merge!

**Comment by olivermrbl** [comment]
Merging this now not to have it hanging too long.

**Pull request merged**